### PR TITLE
[ONNX] Quantization support for five ops

### PR DESCRIPTION
**Pull request by justinchuby**
- Add quantization support for `interpolate`, `avgpool`, `sigmoid` and `add_relu`
- Return the inputs to ListUnpack if the previous node is ListConstruct so that `ListConstruct` and `ListUnpack` are canceled and removed in the jit passes. ONNX doesn't support them.

### CODE REVIEW

**Comment by BowenBao** [comment]
please check CI onnx failure

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -4961,7 +4964,26 @@ def ListConstruct(g, *inputs, **kwargs):
...
+            *inputs: The inputs to the ListUnpack Op.
+            **kwargs: Additional keyword arguments.
+
+        Returns:
+            None if unchanged. Inputs to its previous op if the previous op is
+            ListConstruct.
+        """
+        if not inputs:
+            return None
+        first_input = inputs[0]
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:4981]: `inputs[0]` is shorter and just as clear as `first_input`, so I'd get rid of this variable.

Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py:4981]: Done

**Review by garymm — CHANGES_REQUESTED** [review]
Thanks!
Why `test/onnx/__init__.py`?
If you want it in this PR, please add it to the description.

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -4961,7 +4964,25 @@ def ListConstruct(g, *inputs, **kwargs):
...
+        Args:
+            g: The current graph.
+            *inputs: The inputs to the ListUnpack Op.
+            **kwargs: Additional keyword arguments.
+
+        Returns:
+            None if unchanged. Inputs to its previous op if the previous op is
+            ListConstruct.
+        """
+        if not inputs:
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py]: Removed

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -4961,7 +4964,25 @@ def ListConstruct(g, *inputs, **kwargs):
         return None
 
     @staticmethod
-    def ListUnpack(g, *inputs, **kwargs):
+    def ListUnpack(g, *inputs, **kwargs) -> Optional[List[_C.Value]]:
+        """Translates the ListUnpack Op.
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/symbolic_opset9.py]: Removed

File: test/onnx/symbolic_opsets/test_symbolic_opset9.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+class TestPrim(common_utils.TestCase):
+    def setUp(self):
+        super().setUp()
+        self.graph = _C.Graph()
+
+    def test_list_unpack_returns_all_list_elements_when_previous_node_is_list_construct(
+        self,
+    ):
+        # Build the graph
```
Comment (garymm) [inline test/onnx/symbolic_opsets/test_symbolic_opset9.py:18]: If there's a way to construct this graph via `torch.jit.script` or `torch.jit.trace`, I think that'd be preferable, since it makes it clear what Python code the test applies to.

File: torch/onnx/symbolic_helper.py [hunk 5]
```diff
@@ -287,19 +292,21 @@ def wrapper(g, *args, **kwargs):
             if _zero_point is not None:
                 _zero_point = g.op("Constant", value_t=torch.tensor(_zero_point))
 
-            # some args may be optional, so the length may be smaller
-            assert len(arg_q_descriptors) >= len(args)
-            desc_args = tuple(zip(arg_q_descriptors[: len(args)], args))
+            # Support variable length arguments by marking unspecified ones as non-quantized
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply torch/onnx/symbolic_helper.py:295]: I created #78775. This PR will depend on #78775 since we need the feature to add quantization support for `_interpolate` (which has variable length parameters)

File: test/onnx/symbolic_opsets/test_symbolic_opset9.py (hunk 4 above)
Comment (justinchuby) [reply test/onnx/symbolic_opsets/test_symbolic_opset9.py:18]: My concern is that if we constructed the graph via jit, we make assumptions and rely on the implementation of jit and the various onnx passes. Creating a graph this way removes those assumptions.

Comment (garymm) [reply test/onnx/symbolic_opsets/test_symbolic_opset9.py:18]: Sure, it's OK to keep this as a unit test, but let's also make sure we have an integration test where we go from Python -> jit Graph -> symbolic function.
Do we have any tests in test_pytorch_onnx_onnxruntime.py that would fail without the change to ListUnpack?
If so, then I'm happy.
If not, could you add a test that starts with Python and would fail without the change to ListUnpack?

Comment (justinchuby) [reply test/onnx/symbolic_opsets/test_symbolic_opset9.py:18]: Good point thanks! I created a test.

**Comment by justinchuby** [comment]
> Thanks!
> Why `test/onnx/__init__.py`?
> If you want it in this PR, please add it to the description.

I removed it

**Comment by garymm** [comment]
Newly added test failed:
`FAILED test/onnx/test_pytorch_onnx_onnxruntime.py::TestONNXRuntime_opset10::test_listunpack_runs_without_error_with_constructed_list_as_input`

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 6]
```diff
@@ -3117,27 +3117,66 @@ def test_groupnorm_noaffine(self):
 
     @skipIfUnsupportedMinOpsetVersion(9)
     def test_listunpack(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3119]: can you please rename this to `test_list_unpack`? Generally `CamelCase` -> `snake_case`, not `snakecase`

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 7]
```diff
@@ -3117,27 +3117,66 @@ def test_groupnorm_noaffine(self):
...
+            x,
+            input_names=["x"],
+            dynamic_axes={"x": [0, 1]},
+        )
+        self.run_test(torch.jit.script(ListUnpack()), x, remained_onnx_input_idx=[])
 
-        class ListUnpackSlice(torch.jit.ScriptModule):
-            @torch.jit.script_method
+    @skipIfUnsupportedMinOpsetVersion(9)
+    def test_listunpack_runs_without_error_with_constructed_list_as_input(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3135]: ditto on test name

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 8]
```diff
@@ -3117,27 +3117,66 @@ def test_groupnorm_noaffine(self):
...
+
+            def forward(self, a, b):
+                packed = [a, b]
+                c, _ = packed
+                return c
+
+        self.run_test(
+            torch.jit.script(PackUnpack()),
+            (torch.tensor(0), torch.tensor([42])),
+            input_names=["a", "b"],
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3159]: Why is this and the next arg required?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 9]
```diff
@@ -3117,27 +3117,66 @@ def test_groupnorm_noaffine(self):
...
+            remained_onnx_input_idx=[0],
+        )
+
+    @skipIfUnsupportedMinOpsetVersion(9)
+    def test_listunpack_slice(self):
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3164]: ditto on test name

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 10]
```diff
@@ -12284,13 +12323,14 @@ def forward(self, x):
 
     @skipIfUnsupportedMinOpsetVersion(10)
     def test_qat_avg_pool2d(self):
-        model = torch.nn.Sequential()
-        model.add_module("quant", torch.quantization.QuantStub())
-        model.add_module(
-            "avgpool", torch.nn.AvgPool2d(kernel_size=3, stride=2, padding=1)
+        layers = OrderedDict(
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12326]: I'd in-line this into the constructor call rather than assign to a local. The constructor only takes one argument so it's plenty readable.

Also I don't think naming the modules is really helpful. Can we pass in a tuple or list of modules without names?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 11]
```diff
@@ -12301,13 +12341,14 @@ def test_qat_avg_pool2d(self):
 
     @skipIfUnsupportedMinOpsetVersion(11)
     def test_qat_upsample_nearest2d(self):
-        model = torch.nn.Sequential()
-        model.add_module("quant", torch.quantization.QuantStub())
-        model.add_module(
-            "upsample_nearest2d", torch.nn.UpsamplingNearest2d(scale_factor=1.5)
+        layers = OrderedDict(
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12344]: ditto

**Review by garymm — APPROVED** [review]
Thanks!

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 6 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3119]: There is another `test_list_unpack` which was testing the non scripted version of the model. Is there a convention we use to distinguish scripted and non-scripted tests?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 10 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:12326]: > Also I don't think naming the modules is really helpful. Can we pass in a tuple or list of modules without names?

Good catch. I was working on a parameterized version of the test which needed the module names to fuse/manipulate the modules. This is the remnant of it. I will remove.

> I'd in-line this into the constructor call rather than assign to a local. The constructor only takes one argument so it's plenty readable.

Sure, initially done to remove a level of indent since the model initializations got long

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 6 above)
Comment (garymm) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3119]: Hah! Not that I know of, and if that is a convention we should not continue it. Normally one test covers both. If something can only run in script mode sometimes it is named with `script`, which seems like a much better convention to me.

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 8 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3159]: Thanks for catching. I don't think it is required. The next line is needed because

> Unused inputs will be removed in the exported ONNX model, supplying all inputs will cause an error on unexpected inputs. This parameter tells the verifier which inputs to pass into the ONNX model.

I updated the docstring in verify to clarify its use

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 11 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:12344]: Done

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 10 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:12326]: Done

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 9 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3164]: Done

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 8 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3159]: Done

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 6 above)
Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: Done

**Review by garymm — APPROVED** [review]
🪨
